From 9803d6bf775c6321d2ab7a7d8583865617b71050 Mon Sep 17 00:00:00 2001
From: Damian Hobson-Garcia <dhobsong@igel.co.jp>
Date: Thu, 22 Nov 2012 16:28:27 +0900
Subject: [PATCH 359/859] shmobile: armadillo800eva enable IOMMU support

Add the IPMMU device and register the LCDC, VPU, and VIO devices to allocate
DMA memory via the IPMMU.

Signed-off-by: Damian Hobson-Garcia <dhobsong@igel.co.jp>
---
 arch/arm/mach-shmobile/board-armadillo800eva.c |  3 +++
 arch/arm/mach-shmobile/setup-r8a7740.c         | 23 +++++++++++++++++++++++
 2 files changed, 26 insertions(+)

diff --git a/arch/arm/mach-shmobile/board-armadillo800eva.c b/arch/arm/mach-shmobile/board-armadillo800eva.c
index b4f0224..72690cd 100644
--- a/arch/arm/mach-shmobile/board-armadillo800eva.c
+++ b/arch/arm/mach-shmobile/board-armadillo800eva.c
@@ -40,6 +40,7 @@
 #include <mach/common.h>
 #include <mach/irqs.h>
 #include <mach/r8a7740.h>
+#include <mach/ipmmu.h>
 #include <media/mt9t112.h>
 #include <media/sh_mobile_ceu.h>
 #include <media/soc_camera.h>
@@ -1177,6 +1178,8 @@ static void __init eva_init(void)
 
 	r8a7740_add_standard_devices();
 
+	ipmmu_add_device(&lcdc0_device.dev);
+
 	platform_add_devices(eva_devices,
 			     ARRAY_SIZE(eva_devices));
 
diff --git a/arch/arm/mach-shmobile/setup-r8a7740.c b/arch/arm/mach-shmobile/setup-r8a7740.c
index f3e64e6..4d36909 100644
--- a/arch/arm/mach-shmobile/setup-r8a7740.c
+++ b/arch/arm/mach-shmobile/setup-r8a7740.c
@@ -35,6 +35,7 @@
 #include <mach/r8a7740.h>
 #include <mach/pm-rmobile.h>
 #include <mach/common.h>
+#include <mach/ipmmu.h>
 #include <mach/irqs.h>
 #include <asm/mach-types.h>
 #include <asm/mach/map.h>
@@ -359,6 +360,23 @@ static struct platform_device cmt10_device = {
 	.num_resources	= ARRAY_SIZE(cmt10_resources),
 };
 
+/* IPMMUI (an IPMMU module for ICB/LMB) */
+static struct resource ipmmu_resources[] = {
+	[0] = {
+		.name	= "IPMMUI",
+		.start	= 0xfe951000,
+		.end	= 0xfe9510ff,
+		.flags	= IORESOURCE_MEM,
+	},
+};
+
+static struct platform_device ipmmu_device = {
+	.name           = "ipmmu",
+	.id             = -1,
+	.resource       = ipmmu_resources,
+	.num_resources  = ARRAY_SIZE(ipmmu_resources),
+};
+
 static struct platform_device *r8a7740_early_devices[] __initdata = {
 	&scif0_device,
 	&scif1_device,
@@ -373,6 +391,7 @@ static struct platform_device *r8a7740_early_devices[] __initdata = {
 	&vpu_device,
 	&vpc_device,
 	&vio_device,
+	&ipmmu_device,
 
 };
 
@@ -778,6 +797,10 @@ void __init r8a7740_add_standard_devices(void)
 	/* add devices */
 	platform_add_devices(r8a7740_early_devices,
 			    ARRAY_SIZE(r8a7740_early_devices));
+
+	ipmmu_add_device(&vpu_device.dev);
+	ipmmu_add_device(&vio_device.dev);
+
 	platform_add_devices(r8a7740_late_devices,
 			     ARRAY_SIZE(r8a7740_late_devices));
 
-- 
1.8.1.4

